airflow/api/client/api_client.py

```diff
@@ -70,3 +70,12 @@ def delete_pool(self, name):
         :param name: pool name
         """
         raise NotImplementedError()
+
+    def get_lineage(self, dag_id: str, execution_date: str):
```

Comment:
the `airflow.api.cclient` package is intended for use by CLI. Can we call this method from the CLI?

Reply:
I'm not sure what you mean. The CLI can just use one of the interfaces to the API: local and json. The "local_client" should be eventually deprecated as it touches the database directly and that's a security concern. 

Other application can use this package as well to interface with the API.

Reply:
In my opinion, this class is not intended for external use. Someone can use it, but it requires Airflow installation. I think it would be better if we recommend using the REST API directly. This is described in detail in the documentation - [here](https://airflow.readthedocs.io/en/latest/usage-cli.html#set-up-connection-to-a-remote-airflow-instance) For me, this is an internal abstraction that allows us to execute some CLI commands using the REST API or execute some CLI command using a direct database query.  So if we add a method here, then we should also add the corresponding command in CLI. We should not allow developer to use this class directly because we have a lot of side effects when importing the `airflow` package e.g. changing the logging configuration or importing the webserver classes.

We cannot deprecate local_client because only the CLI that uses local_client is fully functional and we use this class in the default configuration. 

The bit confusing is that this code is in the `api` package and not in the `cli` package, but the `cli` package was created later and it is possible that we should move this code now.

I would like to deal with the subject of API in the near future. I've planned it in Q1. I will present AIP in the near future.

Reply:
The class is intended for external use (I'm the original author). As mentioned the local_client is a security problem and only exists for backwards compatibility. The REST API can be used directly and the json_client provides you a way to do this in any python application which is a provided implementation of the REST API's interface rather than needing to create your own client. 

The CLI should switch to the `json_client` or, in the future, `grpc` maybe. The local_client should be deprecated asap cause there is no way to do proper authn/authz . Basically if you can run the command line client you have full control over Airflow operationally.

You can argue that the client stuff could be moved to a semantic better location but that will not be `cli`.

Reply:
For the AIP I would expect building on top of FAB or equivalent. That gives OpenAPI endpoints out of the box.

Reply:
I think for now it's OK to follow the current "experimental" approach and add this as it is. But we should discuss the approach for API in general soon and define/agree in community both - our target for API implementation and our roadmap how we will get there.

For now I think we have no other choice but implement it this way if this is needed for some other work, but we should be aware it might get deprecated in case we decide that we replace the experimental API with another approach.

Certainly we should not change the approach before we decide where we are going.